downloader/src/main/kotlin/Downloader.kt

```diff
@@ -166,10 +167,11 @@ class Downloader {
             targetDir.safeDeleteRecursively()
             targetDir.safeMkdirs()
 
-            exception.addSuppressed(e)
+            e.initCause(lastCause)
```

Comment:
Independent from the above I believe one can also argue `addSuppressed` was a better fit because your change actually turns a `tree` into a `list` IIUC. While this functions calls two child actions and only if `both` of them fail it `throws`. From that perspective on could argue that neither the `first` child action can be the cause of the `second` nor the other way round as they are `siblings`. I fear I didn't manage to write it clear, does it make some sense to you ?

Reply:
It's more like the opposite: Suppressed exceptions are stored in a flat array, and there is no order implied, e.g. that the exception that was suppressed first also comes first in the array. Using causes, you have an implied tree structure because each exception can only have a single cause, so the cause of the cause is stored in the exception one level down the tree.

Reply:
Hm, let me re-phrase to try to avoid misunderstandings. I understand that this change make the tree look like this:
```
download:
- vcsDownload 
  - sourceArtifactDownload
```
While before it looked like:
```
download:
- vcsDownload 
- sourceArtifactDownload
```

Reply:
Actually, the vcsDownloadException is the leaf (as that's what we try first, so it's also the first the could fail):

    DownloadException
        sourceDownloadException
            vcsDownloadException 

But I start to get your point. (Will edit this further)

Reply:
How would your desired output look like?

Reply:
I'd like to have the last two lines (in this example) swapped, so that the exception that occurred / was suppressed first comes last, as in a usual Java stack trace:

```
DownloadException: Download failed for 'Gradle:com.github.ajalt:source:2.5.0-beta1'.
Suppressed: DownloadException: No source artifact URL provided for 'Gradle:com.github.ajalt:source:2.5.0-beta1'.
Suppressed: DownloadException: Unable to determine a revision to checkout.,
```

Reply:
Hm, it might be confusing for people familiar with the default way the `suppressed` exceptions are printed. If one reads `suppressed` as I tree I personally do not see why order matters at all.
This said just as a note or my 2 cents.

Reply:
> If one reads `suppressed` as I tree I personally do not see why order matters at all.

Order matters to me as it gives an indication about what was tried first before falling back to other code paths.

But I guess we'll not reach a consensus here. I'll close unmerged.

Reply:
I will neither approve nor block this, so up to you.